src/nvim/decoration.c

```diff
@@ -191,6 +192,16 @@ void buf_put_decor_sh(buf_T *buf, DecorSignHighlight *sh, int row1, int row2)
     sh->sign_add_id = sign_add_id++;
     if (sh->text[0]) {
       buf_signcols_count_range(buf, row1, row2, 1, kFalse);
+      // Since number_width() is a cached value, it may not be updated accordingly after
+      // adding the first sign so do it here in case it's smaller than 2 and 'signcolumn'
+      // is set to "number" #28984.
+      if (buf_meta_total(buf, kMTMetaSignText) == 1) {
+        FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
```

Comment:
Should this also be done for other tabpages?

Reply:
It seems so. I assumed the nrwidth cache would be invalidated upon switching tabpages but that doesn't seem to be the case.

There is also (currently `static`) `may_force_numberwidth_recompute()` which could be re-used here.

Reply:
Updated with a different solution that handles this properly.